src/SharedStream.mo

```diff
@@ -0,0 +1,380 @@
+      };
+      if (firstIndex != expectedNextIndex_) {
+        Debug.trap("Broken chunk index: " # Nat.toText(firstIndex) # "; expected: " # Nat.toText(expectedNextIndex_));
+      };
+      for (index in chunk.keys()) {
+        callback(streamId, chunk[index], firstIndex + index);
+      };
+      expectedNextIndex_ += chunk.size();
+      #ok();
+    };
+
+  };
+
+  /// Usage:
+  ///
+  /// let sender = StreamSender<Int>(
+  ///   123,
+  ///   10,
+  ///   10,
+  ///   func (item) = 1,
+  ///   anotherCanister.appendStream,
+  /// );
+  /// sender.next(1);
+  /// sender.next(2);
+  /// .....
+  /// sender.next(12);
+  /// await* sender.sendChunk(); // will send (123, [1..10], 0) to `anotherCanister`
+  /// await* sender.sendChunk(); // will send (123, [11..12], 10) to `anotherCanister`
+  /// await* sender.sendChunk(); // will do nothing, stream clean
+  public class StreamSender<T>(
+    streamId : Nat,
+    maxSize : ?Nat,
+    weightLimit : Nat,
+    weightFunc : (item : T) -> Nat,
+    callback : (streamId : Nat, items : [T], firstIndex : Nat) -> async R.Result<(), StreamError>,
+  ) {
+
+    let queue : QueueBuffer.QueueBuffer<T> = QueueBuffer.QueueBuffer<T>();
+    // a head of queue before submitting lately failed chunk. Used for error-handling
+    var lowestError : { #Inf; #Val : Nat } = #Inf;
+
+    public func fullAmount() : Nat = queue.fullSize();
+    public func queuedAmount() : Nat = queue.queueSize();
+    public func nextIndex() : Nat = queue.nextIndex();
+
+    public func get(index : Nat) : ?T = queue.get(index);
+
+    var lastChunkTimestamp : Time.Time = Time.now();
+    var heartbeatTimer : Nat = 0;
+    heartbeatTimer := Timer.recurringTimer(
+      #seconds 30,
+      func() : async () {
+        // if last call was more than 20 seconds ago, send empty chunk
+        if ((Time.now() - lastChunkTimestamp) > 20_000_000_000) {
+          try {
+            switch (await callback(streamId, [], queue.headIndex())) {
+              case (#ok) lastChunkTimestamp := Time.now();
+              case (#err err) Timer.cancelTimer(heartbeatTimer);
+            };
+          } catch (err) {
+            // pass
+          };
+        };
+      },
+    );
+
+    public func next(item : T) : { #ok : Nat; #err : { #NoSpace } } {
+      switch (maxSize) {
+        case (?max) if (queue.queueSize() >= max) {
+          return #err(#NoSpace);
+        };
+        case (_) {};
+      };
+      #ok(queue.push(item));
+    };
+
+    var concurrentChunksCounter : Nat = 0;
+    public func sendChunk() : async* {
+      #ok : Nat;
+      #err : { #Paused; #Busy; #SendChunkError : Text; #StreamClosed };
+    } {
+      if (concurrentChunksCounter >= 5) {
+        return #err(#Busy);
+      };
+      switch (lowestError) {
+        case (#Inf) {};
+        case (#Val _) { return #err(#Paused) };
+      };
+      let headId = queue.headIndex();
+      streamIter.reset();
+      let elements = Iter.toArray(streamIter);
```

Comment:
Iter.toArray is inefficient in that it creates a Buffer first (because it doens't know the length) and then converts the Buffer to an Array.

Is there a better way? We can use `get` to inspect the elements in the queue and so determine the length. Once we know the length we can use `Array.tabulate`.

Or, if we don't want to use `get` from outside, then we have to generalize the `pop` function of QueueBuffer (to return an array and to accept a weight function and a weight limit).

Reply:
What do you think about creating single-linked list from iterable with tracking amount, and then create array with Array.tabulate, where item function will look something like `{ let item = list.0; list := list.1; item; }`? Should be efficient enough without affecting queue

Reply:
Why shall we do anything the allocates memory? If all we need is look at the elements (read) and count then in memory we only need a counter, we don't need to create new data structures.